### PR TITLE
[NBS Docs] Migrate bitbucket server provider to new backend system

### PR DESCRIPTION
**Pull request by camilaibs**
## Hey, I just made a Pull Request!

Closes: https://github.com/backstage/backstage/issues/26403

Migrate the `Bitbucket Server` to a new auth module and update its documentation to use the new Backend System.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/auth-backend/api-report.md
Comment (camilaibs) [inline plugins/auth-backend/api-report.md:436]: Is it fine?

Comment (freben) [reply plugins/auth-backend/api-report.md:436]: No, the types are actually different.

Comment (camilaibs) [reply plugins/auth-backend/api-report.md]: Fixed 🙌🏻

File: plugins/auth-backend-module-bitbucket-server-provider/src/authenticator.ts
Comment (camilaibs) [inline plugins/auth-backend-module-bitbucket-server-provider/src/authenticator.ts:26]: It looks like the profile was fetched in a specific way before, should it be replicated here in this transform? The legacy fetch profile function had access to the config, how can I access it here?

File: plugins/auth-backend-module-bitbucket-server-provider/src/authenticator.ts
Comment (camilaibs) [inline plugins/auth-backend-module-bitbucket-server-provider/src/authenticator.ts:29]: Also could not find any required scopes on the legacy profile logic, can I just remove this?

**Comment by freben** [comment]
needs a fix

File: plugins/auth-backend-module-bitbucket-server-provider/src/authenticator.ts
Comment (camilaibs) [reply plugins/auth-backend-module-bitbucket-server-provider/src/authenticator.ts:94]: I updated the logic to fetch the profile manually, please take a look.

File: docs/auth/bitbucketServer/provider.md
Comment (awanlin) [inline docs/auth/bitbucketServer/provider.md:49]: We need to add the sign in resolvers config to the section above. We also have on all the auth providers a Resolvers sub-section below like this one: https://backstage.io/docs/auth/bitbucket/provider#resolvers

File: docs/auth/bitbucketServer/provider.md
Comment (awanlin) [inline docs/auth/bitbucketServer/provider.md:54]:
```suggestion
yarn --cwd packages/backend add @backstage/plugin-auth-backend-module-bitbucket-server-provider
```

File: docs/auth/bitbucketServer/provider.md
Comment (awanlin) [inline docs/auth/bitbucketServer/provider.md:59]: We can simplify this to just be the new `backend.add()` like we have done here: https://backstage.io/docs/auth/bitbucket/provider#backend-installation

File: plugins/auth-backend/src/providers/bitbucketServer/provider.ts
Comment (awanlin) [inline plugins/auth-backend/src/providers/bitbucketServer/provider.ts:1]: I don't think we deleted this for the other auth providers, would this then be a breaking change?

**Review by awanlin — CHANGES_REQUESTED** [review]
Thanks for running with this @camilaibs, over all looks good, I'm not sure about the auth backend changes though? Do we need those or should we remove that all at once in a future PR that includes the other auth providers?

File: plugins/auth-backend/src/providers/bitbucketServer/provider.ts
Comment (camilaibs) [reply plugins/auth-backend/src/providers/bitbucketServer/provider.ts:1]: As recommended [here](https://github.com/backstage/backstage/issues/19476), I used the [gitlab provider pull request](https://github.com/backstage/backstage/pull/19475) as an example, and there was similar deletion in the changes.

**Comment by camilaibs** [comment]
> needs a fix

Ready for re-review @freben 🙂

**Comment by knowacki23** [comment]
Hey,
stupid question - will the new module `auth-backend-module-bitbucket-server-provider` be released and ready to use right after this PR gets merged or we would have to wait for new Backstage release? :pray:

**Comment by camilaibs** [comment]
> Hey, stupid question - will the new module `auth-backend-module-bitbucket-server-provider` be released and ready to use right after this PR gets merged or we would have to wait for new Backstage release? 🙏

Hi @knowacki23 we follow these [release lines](https://backstage.io/docs/overview/versioning-policy/#release-lines), so when a pull request gets merged it will be available in the upcoming **next** and **main** releases 🙂 .

File: plugins/auth-backend-module-bitbucket-server-provider/src/authenticator.ts
Comment (camilaibs) [reply plugins/auth-backend-module-bitbucket-server-provider/src/authenticator.ts]: Removed for now, please let me know if there is any required scope I should define here 🙂

**Comment by knowacki23** [comment]
> > Hey, stupid question - will the new module `auth-backend-module-bitbucket-server-provider` be released and ready to use right after this PR gets merged or we would have to wait for new Backstage release? 🙏
> 
> Hi @knowacki23 we follow these [release lines](https://backstage.io/docs/overview/versioning-policy/#release-lines), so when a pull request gets merged it will be available in the upcoming **next** and **main** releases 🙂 .

Thank you for the fast reply! :slightly_smiling_face:

**Review by freben — APPROVED** [review]
Yep looks great!

File: plugins/auth-backend-module-bitbucket-server-provider/src/helpers.ts
Comment (dhruvaldarji) [inline plugins/auth-backend-module-bitbucket-server-provider/src/helpers.ts:24]: Note that Bitbucket servers allow for the configuration of custom Api Base URLs. 
This means we cannot assume that the server address can be replicated using only a host address. Instead, we would need the entire `apiBaseUrl`. 

For example, a Bitbucket server address like "https://bitbucket.company-domain.com" is possible, but it's also possible to have a server address like "https://company-domain.com/bitbucket." 

However, fewer companies employ the latter strategy, with [Atlassian dropping support for Bitbucket Server](https://bitbucket.org/blog/cloud-migration-benefits). My organization has even opted to migrate to a different service provider rather than continue to use Bitbucket Server, with its caveats of custom API Base URLs.

File: plugins/auth-backend-module-bitbucket-server-provider/src/helpers.ts
Comment (dhruvaldarji) [inline plugins/auth-backend-module-bitbucket-server-provider/src/helpers.ts:41]: The `X-Ausername` header may not always conform to the value expected by the Bitbucket Server REST API to designate the user. You may have to massage the data here to convert `@` to `_`.

bitbucket-server-whoami-headers

Organizations that populate their users in Bitbucket using SAML/LDAP providers often designate usernames via email addresses. However, the Bitbucket username API does not accept usernames with `@` symbols. As such, while the display name may contain the symbol, the REST API requires that `@` symbols be replaced with `_`. 

For example, a user with the username of `ddarji@company.com` would have the username of `ddarji_company.com.` 

bitbucket-server-user-response
bitbucket-server-user-headers

File: plugins/auth-backend-module-bitbucket-server-provider/src/helpers.ts
Comment (camilaibs) [reply plugins/auth-backend-module-bitbucket-server-provider/src/helpers.ts:24]: Hi @dhruvaldarji, thanks a lot for your feedback, I really appreciate that 🙏🏻. Just to confirm, you are suggesting to rename the configuration to `baseUrl`, right? Since in this pull request I was just porting the legacy provider code to the new auth module format, I wonder if it would be a good idea to make the changes you suggested in a follow-up pull request. Let's ask @freben what he suggests for us 🙂

File: plugins/auth-backend-module-bitbucket-server-provider/src/helpers.ts
Comment (camilaibs) [reply plugins/auth-backend-module-bitbucket-server-provider/src/helpers.ts:41]: That's a good catch, thanks for pointing out 🙌🏻 . Do you think this transformation should always be done or it could be optionally enabled via config? As I mentioned in the comment above, I'm just porting the legacy provider code to the new auth module format, I wonder if it would be a good idea to make the changes you suggested in a follow-up pull request. Let's see what @freben says 🙏🏻

File: plugins/auth-backend-module-bitbucket-server-provider/src/helpers.ts
Comment (dhruvaldarji) [reply plugins/auth-backend-module-bitbucket-server-provider/src/helpers.ts:24]: No changes requested. I'm sure this will be fine. The user base is dwindling anyway, since Atlassian dropped support.

File: plugins/auth-backend-module-bitbucket-server-provider/src/helpers.ts
Comment (dhruvaldarji) [reply plugins/auth-backend-module-bitbucket-server-provider/src/helpers.ts:41]: I believe the conversion should always be done, but as you said, it can be done in a separate PR. If someone really needs it, they will propose the change. 

Normally, that someone would have been me, but my organization is no longer using this provider, so there's no reason to hold this back. 😅

**Review by dhruvaldarji — APPROVED** [review]
As noted in the comments, Atlassian has dropped support for Bitbucket Server. As such, many organizations are shifting to other providers. This reduces the priority on complete support for every feature made available through the Bitbucket Server. 

The changes for supporting Bitbucket Server API Base URLs and for supporting  organizations that use emails for usernames can be done at a later time, if they are so desired by existing users. 

With that in mind, this changeset is good to go. 👍

File: plugins/auth-backend-module-bitbucket-server-provider/api-report.md
Comment (Rugvip) [inline plugins/auth-backend-module-bitbucket-server-provider/api-report.md:41]: This doesn't match the authenticator itself, which ends up using `OAuthAuthenticatorResult<PassportProfile>`. The `BitbucketServerOAuthResult` type also contains the same information as the new `OAuthAuthenticatorResult` so I think we should be able to get rid of it.

**Review by Rugvip — COMMENTED** [review]
Nice! 👍 

Just the session type bits, let's indeed leave the profile fetch fixes for potential followup PRs.

**Review by Rugvip — APPROVED** [review]
Nice! 👍 :shipit:

File: plugins/auth-backend-module-bitbucket-server-provider/api-report.md
Comment (camilaibs) [reply plugins/auth-backend-module-bitbucket-server-provider/api-report.md]: Fixed 🙏🏻

File: plugins/auth-backend/src/providers/bitbucketServer/provider.ts
Comment (awanlin) [reply plugins/auth-backend/src/providers/bitbucketServer/provider.ts:1]: All good with me 👍

**Review by awanlin — APPROVED** [review]

**Pull request merged**
